Netcore: Migrated remaining backoffice controllers

bergmania opened the pull request:
# Notes
- Updated generated templates to use the current namespace of `UmbracoViewPage` and `PartialViewMacroPage`
- Fixed some issues with content-root vs web-root
- Fixed some test builders
- Migrated TestServer tests (Explicit until we have a solution for build server)
- Added AutoFixture and created an `AutoMoqDataAttribute` in the unit test project, to simplify tests of UsersController and can be used in the future
- Fixed issues when services injected `IPasswordConfiguration`. Which is not registered in the container. Instead, I inject `IUserPasswordConfiguration`
- Fixed issue in `DetermineAmbiguousActionByPassingParameters` when null-values are valid.
- Added filter to output exceptions as JSON on `UmbracoAuthorizedJsonController`
- Added new options to `LinkGeneratorExtensions` to generate links with multiple query params
- Fixed issue with the `UmbracoJsonModelBinder` so now `[Required]` attributes are not required to model bind
   - The issue here is e.g. `UserSave` is not model bound and controller throws ``ArgumentNullException`.
- Migrated the following controllers:
    - `CurrentUserController`
    - `MemberGroupController`
    - `TemplateQueryController`
    - `UserGroupsController`
    - `UsersController`

- Migrated the following actions into existing controllers
    - `AuthenticationController.GetPasswordConfig`

- Finally also removed a lot of files that is not used anymore in the `Umbraco.Web` project

### Questions
- How can we test FIPS compliance for users? Old tests commented out (`UsersControllerTests`)

### Tests
- In general, BackOffice should work. Espespecially error/validation messages.
- Known issues
   - The Cultures set on the thread is not chosen from the user signing in but is using the machine default (I have danish 👎 )
   - The users are not updated on login, so the user is set marked as inactive

stevetemple commented on src/Umbraco.Web.Common/Formatters/IgnoreRequiredAttributsResolver.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace Umbraco.Web.Common.Formatters
+{
+    public class IgnoreRequiredAttributsResolver : DefaultContractResolver
```
```suggestion
    public class IgnoreRequiredAttributesResolver : DefaultContractResolver
```

stevetemple commented on src/Umbraco.Tests.Common/Builders/UserBuilder.cs:
```diff
@@ -76,12 +79,20 @@ public UserBuilder<TParent> WithLanguage(string language)
             return this;
         }
 
+   public UserBuilder<TParent> WithUsername(string username)
```
Formatting: spacing

stevetemple commented on src/Umbraco.Web.Common/ModelBinding/UmbracoJsonModelBinderProvider.cs:
```diff
@@ -25,6 +29,11 @@ private static IInputFormatter[] GetNewtonsoftJsonFormatter(ILoggerFactory logge
+            var ss = jsonOptions.SerializerSettings; // Just use the defaults as base
+
+            // We need to ignore required attributes when serializing. E.g UserSave.ChangePassword. Otherwise the model is not model bound.
+            ss.ContractResolver = new IgnoreRequiredAttributsResolver();
```
```suggestion
            ss.ContractResolver = new IgnoreRequiredAttributesResolver();
```

stevetemple commented on src/Umbraco.Web.BackOffice/Filters/UserGroupValidateAttribute.cs:
```diff
@@ -0,0 +1,101 @@
+                context.ModelState.AddModelError("Alias", "A user group with this alias already exists");
+            }
+
+            // TODO: Validate the name is unique?
```
TODO

bergmania replied:
Not a new TODO.. Just copied over from v8

stevetemple commented on src/Umbraco.Web.BackOffice/Filters/UserGroupValidateAttribute.cs:
```diff
@@ -0,0 +1,101 @@
+                        return;
+                    }
+
+                    //map the model to the persisted instance
```
These comments here and just below are more what rather than why

bergmania replied:
Agree.. But again, these are just copied over one to one

stevetemple commented on src/Umbraco.Web.BackOffice/Filters/UserGroupValidateAttribute.cs:
```diff
@@ -0,0 +1,101 @@
+        {
+            _userService = userService ?? throw new ArgumentNullException(nameof(userService));
+            _umbracoMapper = umbracoMapper ?? throw new ArgumentNullException(nameof(umbracoMapper));
+        }
```
Formatting: Spacing

stevetemple commented on src/Umbraco.Web.BackOffice/Security/PasswordChanger.cs:
```diff
@@ -0,0 +1,102 @@
+
+            if (passwordModel.NewPassword.IsNullOrWhiteSpace())
+            {
+                return Attempt.Fail(new PasswordChangedModel { ChangeError = new ValidationResult("Cannot set an empty password", new[] { "value" }) });
```
Can these strings be localised?

bergmania replied:
I think yes, but not really related to this PR. This is just copied over. We could make an up-for-grabs task about localizing them

stevetemple commented on src/Umbraco.Web/Editors/PasswordChanger.cs:
```diff
@@ -9,6 +9,7 @@
 using Umbraco.Web.Security;
 using IUser = Umbraco.Core.Models.Membership.IUser;
 
+//Migrated to .NET CORE
```
Can this be deleted

bergmania replied:
Nope, not yet.. Still used in members implementation

stevetemple commented on src/Umbraco.Tests.Common/Builders/UserBuilder.cs:
```diff
@@ -113,7 +124,7 @@ public UserBuilder<TParent> WithSuffix(string suffix)
         public override User Build()
         {
-            var id = _id ?? 1;
+            var id = _id ?? 0;
```
This feels intentional but possibly breaking, not sure why it was 1 previously but thought I'd raise in case it's accidental

bergmania replied:
Yes.. But this is just the Test builder. I think it was by mistake 1 before.. All tests updated

bergmania commented:
Thanks, @stevetemple..

Good feedback. I think all of the concerns are relevant, but not necessarily for this PR, as all the issues existed before.

I fixed the misspelling and formatting issues.

emmagarland commented on src/Umbraco.Web.Common/Formatters/IgnoreRequiredAttributesResolver.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace Umbraco.Web.Common.Formatters
+{
+    public class IgnoreRequiredAttributesResolver : DefaultContractResolver
```
Can we create tests for this new resolver?

bergmania replied:
I added one :)

emmagarland commented on src/Umbraco.Web.BackOffice/Controllers/MemberGroupController.cs:
```diff
@@ -0,0 +1,106 @@
+        }
+
+        [HttpDelete]
+        [HttpPost]
```
I know this was already here, just wondered why we might have POST here too

bergmania replied:
I'm not really sure, but at least the client are using POST. Whether delete is used, I don't know.

emmagarland commented on src/Umbraco.Web/Runtime/WebInitialComposer.cs:
```diff
@@ -31,7 +31,6 @@ public override void Compose(Composition composition)
 
             composition.Register<UmbracoInjectedModule>();
 
-            composition.ComposeWebMappingProfiles();
```
I don't know what the implications of this will be since it has been deleted, but are we registering all the previous components somewhere else now (or future task)?

bergmania replied:
I moved some of the registrations to the `ComposeCoreMappingProfiles`, and the rest is added in `BackOfficeComposer`

emmagarland commented on src/Umbraco.Web.BackOffice/Controllers/UsersController.cs:
```diff
@@ -147,30 +172,24 @@ internal static async Task<HttpResponseMessage> PostSetAvatarInternal(HttpReques
                 userService.Save(user);
-
-                //track the temp file so the cleanup filter removes it
-                tempFiles.UploadedFiles.Add(new ContentPropertyFile
```
Was this not needed then?

bergmania replied:
It is not needed anymore. The reason is because in framework all files were uploaded to a temp location before the controller action was executed. Therefore we had to clean up (using the `FileUploadCleanupFilterAttribute`).

This is not the case anymore, and we handle the stream directly in the action, so nothing to clean up,

emmagarland commented:
Awesome to have the .NET Core Umbraco back-office running locally! 🥇 

Will put any issues found so far on here. Let it be known I have break on **_all_** exceptions switched on!! Some of these maybe known or wider issues we can break out.

**Testing Results so far (only just dipped into the back-office atm)**

1. I usually default to SQLCE and not SQLServer when testing/running locally, but this didn't work as it needs a SQLCE project reference to do so. However, SQLCE being the default Umbraco DB is usual behaviour for the current v8 install. As you prob have thought of some people who run Umbraco for dev may not have full SQL or just need something spun up quickly, so perhaps will expect the same as I did by default, like previous Umbraco versions.
2. When I customise the install and choose a brand new database, I see some `SqlServerSyntaxProvider.cs` exceptions being thrown when initially creating the new SQLServer database. Things like `GetColumnsInSchema` method throwing an exception because `TABLE_NAME`, `COLUMN_NAME`  doesn't exist etc. I believe this was line 190 `GetColumnsInSchema`
3. All the database tables were created, along with my default user, which was cool. However, then I seemed to get into some sort of install loop. The page kept saying _"You see this screen because your Umbraco installation did not complete correctly"_. I tried to debug why but it just went through some checks in `NewInstallStep.cs` and then kept returning that view `continueinstall.cshtml`. I ended up stopping the site, clearing cookies, rerunning the app and logging in with the credentials I'd used - I logged in fine. Now I am in the .NET Core Umbraco back-office! Woohoo!
4. When logging in the first time, there were a few errors being thrown behind the scenes, but they didn't break me out of the login process so I temporarily passed over them to focus on the goal to look at the back-office - we can look at them later. Some might already exist - an example is an exception on an index not being present yet. 
5. While debugging with break all exceptions on, BackgroundTaskRunner kicks in with:

`System.IO.DirectoryNotFoundException: 'Could not find a part of the path 'D:\Source\Repos\Umbraco-v8\src\Umbraco-CMS\src\Umbraco.Web.UI.NetCore\wwwroot\App_Data\TEMP\FileUploads'.' from BackgroundTaskRunner`


Thanks

bergmania commented on src/Umbraco.Tests/TestHelpers/ControllerTesting/TestStartup.cs:
```diff
@@ -39,7 +40,7 @@ public void Configuration(IAppBuilder app)
             httpConfig.Services.Add(typeof (IExceptionLogger), new TraceExceptionLogger());
 
-            httpConfig.Services.Replace(typeof (IAssembliesResolver), new SpecificAssemblyResolver(new[] { typeof (UsersController).Assembly }));
+            httpConfig.Services.Replace(typeof (IAssembliesResolver), new SpecificAssemblyResolver(new[] { typeof (AspNetApplicationShutdownRegistry).Assembly }));
```
We just need the assembly of `Umbraco.Web`. `UsersController` is moved away from  `Umbraco.Web`, so needed just a new class from  `Umbraco.Web`.

bergmania commented:
Thanks, @emmagarland ..

I have added the SqlCE project reference to the netcore executable so now you can use SqlCE as default (on windows). The issue now is when you try from non-windows, this will end up failing. That's a problem for later.

I can't repro the install loop. I tried to install using LocalDb and SqlCE - Both succeed.

Regarding the review, I added the suggested tests and answered on the comments.

Shazwazza commented on src/Umbraco.Tests.UnitTests/Umbraco.Web.Common/Formatters/IgnoreRequiredAttributesResolverUnitTests.cs:
```diff
@@ -0,0 +1,41 @@
+        [DataContract(Name = "objectWithRequiredProperty", Namespace = "")]
+        private class ObjectWithRequiredProperty
+        {
+            [DataMember(Name = "property", IsRequired = true)]
```
So am I correct in assuming that this new IgnoreRequiredAttributesResolver is specifically for deserializing JSON so it actually deserializes without an exception even when IsRequired is there *but* the IsRequired will still be used during validation? Or is the validation Required attribute different? If so, is there ever a need to use the IsRequired param?

bergmania replied:
Yes..
The `RequiredAttribute` is different. I guess that is just used invalidation. 

Maybe we could just remove the `DataMember(IsRequired = true)`. I assumed it was there because we require it hen we output the same model.

Shazwazza replied:
maybe, i think we could revisit it, maybe for now we'll just leave this as-is. seems strange that it worked in v8, or maybe that's only because we switched everything to DataMember stuff in netcore and we didn't have that setting?

bergmania replied:
I think it is because we have our own UmbracoJsonModelBinderConvention now..

Shazwazza commented on src/Umbraco.Tests/Web/Controllers/UsersControllerTests.cs:
```diff
@@ -1,526 +1,526 @@
-        }
-    }
-}
+﻿// using System;
```
Remove this file?

Shazwazza commented on src/Umbraco.Web.BackOffice/Filters/UserGroupValidateAttribute.cs:
```diff
@@ -0,0 +1,101 @@
+
+        }
+
+        }
```
Is this an extra } ? don't think the build is failing but seems odd

bergmania replied:
I will reformat entire fill to fix up :)

Shazwazza commented on src/Umbraco.Web.UI.NetCore/Umbraco.Web.UI.NetCore.csproj:
```diff
@@ -61,6 +62,23 @@
           <SubType>Designer</SubType>
         </Content>
         <Content Remove="wwwroot\Web.config" />
+        <Content Remove="Umbraco\PartialViewMacros\Templates\Gallery.cshtml" />
```
Do we need all these remove statements?

bergmania replied:
Nope.. I will clean up in a later PR regarding the publish for alpha

Shazwazza commented on src/Umbraco.Web.UI.NetCore/Umbraco/PartialViewMacros/Templates/Breadcrumb.cshtml:
```diff
@@ -18,7 +18,7 @@
         @foreach (var item in selection.OrderBy(x => x.Level))
         {
-            <li><a href="@item.Url()">@item.Name</a> <span class="divider">/</span></li>
+            <li><a href="@item.Url(PublishedUrlProvider)">@item.Name</a> <span class="divider">/</span></li>
```
This will be an interesting change for people if we make this a requirement. I'm all for it since the model shouldn't be responsible for generating the URL but I think if we make the change we shouldn't use ext methods on the model to do it and instead just call the service. But we can discuss later.

bergmania replied:
I guess we will end up having an extra nuget package. with the old codesmells. So having the old extention methods available. but we should not use these our self, or in templates.. Code is untestable using these..

`Umbraco.Friendly` or similar sounds good ;)
At least better than `Umbraco.ReintroduceCodeSmells` or `Umbraco.ReintroduceUntestableCode` :P

Shazwazza replied:
Yep we'll see how it goes, i think the more we move away from the hacks the better and then can re-implement 'friendly' things later if necessary. In some cases though i think we'll realize that some of the 'friendly' things don't actually make much sense. For example, when (one day) we decouple the content cache from URLs, then item.Url() probably won't make sense anymore.

bergmania commented on src/Umbraco.Tests.Common/Builders/UserGroupBuilder.cs:
```diff
@@ -92,7 +92,7 @@ public IReadOnlyUserGroup BuildReadOnly(IUserGroup userGroup)
         public override IUserGroup Build()
         {
-            var id = _id ?? 1;
+            var id = _id ?? 0;
```
0 is intended.

This is the userGroup builder... But for user builder 0 is also intended, so it can be created by default. If you need an admin, then you can use WithId(Constants.Security.SuperUserId)

bergmania commented on src/Umbraco.Infrastructure/Scheduling/BackgroundTaskRunner.cs:
```diff
@@ -137,17 +137,17 @@ public BackgroundTaskRunner(BackgroundTaskRunnerOptions options, ILogger logger,
         /// <param name="options">The set of options.</param>
         /// <param name="logger">A logger.</param>
-        /// <param name="hostingEnvironment">The hosting environment</param>
+        /// <param name="applicationShutdownRegistry">The hosting environment</param>
```
Fixing.. But not really sure what value these param-xml docs are given in 99% of the cases, we just humanize the variable name.

Shazwazza replied:
i know, it's just they get used to generate the csharp docs https://our.umbraco.com/apidocs/v8/csharp/api/Umbraco.Core.html
